Assignment-6

smxlikith opened the pull request:
This pull request introduces a complete implementation of a Virtual File System (VFS) in C, simulating an in-memory file storage environment using circular and doubly linked lists. The system supports typical file and directory operations, manages free disk blocks, and provides a command-line interface for user interaction. The changes span multiple new files, each responsible for a core component of the VFS, including memory management, file system structure, user input parsing, command handling, and the main application logic.

The most important changes are:

**Core File System Structure and Operations:**

* Introduced the `FileNode` structure and associated functions in `fileSystem.c`/`fileSystem.h` to represent files and directories, support path resolution, and manage hierarchical relationships using circular linked lists. 
* Implemented file and directory operations (create, delete, mkdir, rmdir, cd, ls, pwd, read, write, df) in `commands.c`/`commands.h`, providing the main interface for manipulating the VFS.

**Memory and Block Management:**

* Added `Block` structure and memory block management functions in `memBlock.c`/`memBlock.h`, including initialization, allocation, deallocation, and freeing of blocks using a doubly linked list to simulate disk space. 

**User Input and Command Parsing:**

* Developed robust user input handling in `userInput.c`/`userInput.h`, including parsing commands and arguments (with quote support), and safe memory management for command-line interaction.

**Main Application and Command Loop:**

* Created `virtualFileSystem.c` as the entry point, handling initialization, command dispatch, and the main event loop for the VFS, integrating all subsystems and providing error handling and cleanup.

**Documentation:**

* Added a comprehensive `README.md` describing the VFS features, supported operations, compilation instructions, and implementation details.

smxlikith committed:
Completed Assignment-6

tejas0123 commented on Assignment-6/lib/commands.c:
```diff
@@ -0,0 +1,192 @@
+
+    if (target == root) {
+        printf("Error: cannot remove root '/'\n", path);
+        return;
```
Wouldn't the isDirectory check suffice for root ?

smxlikith replied:
The root refers to the root of the entire file system. The `isDirectory` attribute tells us whether the target `FileNode` is a directory. For example:

```bash
rmdir "/temp/temp.txt"
```

In this example, the target `FileNode` is a file (`temp.txt`), so the `isDirectory` attribute would indicate that it's not a directory.

```bash
rmdir "/"
```

In this case, the target `FileNode` is the root directory itself. An additional check would be required in this scenario to handle the root directory properly.

smxlikith replied:
I wanted to give a different error since it's the root of the file system, but yes, just the isDirectory check would suffice.

tejas0123 commented on Assignment-6/virtualFileSystem.c:
```diff
@@ -0,0 +1,99 @@
+    } else if (strcmp(args[0], "exit") == 0) {
+        printf("Memory released. Exiting program...\n");
+        freeBlocks(*freeBlocksList);
+        free(virtualDisk);
```
What about filenodes ?

tejas0123 replied:
They should be cleaned up too

smxlikith replied:
Yes, I will be adding that in the next commit.

smxlikith committed:
Fixed Exit function

tejas0123 commented on Assignment-6/lib/commands.c:
```diff
@@ -0,0 +1,192 @@
+    if (!target->isDirectory) {
+        printf("Error: failed to remove '%s': Not a directory\n", path);
+        return;
+    }
```
Combine some of these validations with the validations in the deleteFile function and make it reusable. It should basically validate a fileNode for deletion. It can take the fileNode type (dir or file) as another input

tejas0123 commented on Assignment-6/lib/userInput.c:
```diff
@@ -0,0 +1,92 @@
+    args[currArgs - 1] = NULL;
+    *size = currArgs - 1;
+    free(arg);
+    return args;
```
- Remove the tokens and use strtok to tokenise
- Use a fixed size array of some max size and remove rellocation logic
- This function is doing too mucj, so break it according to the functionality. Helper functions need not be in the header file

tejas0123 commented on Assignment-6/lib/commands.c:
```diff
@@ -0,0 +1,192 @@
+    }
+}
+
+void writeFile(FileNode* root, FileNode* cwd, Block** freeBlocksList, char* path, char* data) {
```
Too many arguments. See if you can create a struct variable for this with the properties that the write function takes

tejas0123 replied:
Do this other functions too if you find it relevant and if you can group the arguments into a single logical entity

tejas0123 commented on Assignment-6/lib/commands.c:
```diff
@@ -0,0 +1,192 @@
+        file->blocks = allocateBlocks(freeBlocksList, n);
+    }
+    Block* block = file->blocks;
+    for (int i = 0; i < n; i += BLOCK_SIZE) {
+        size_t copySize = strlen(data + i) > BLOCK_SIZE ? BLOCK_SIZE : strlen(data + i);
+        strncpy(block->data + block->currSize, data + i, copySize);
+        block->currSize += copySize;
+        block = block->next;
```
Good logic

smxlikith committed:
Made changes:
	- added FileNode Validater.
	- added VFSObject to group function Arguments.
	- changed userInut to use static allocation.

tejas0123 reviewed (approved):
LGTM
